### ISSUE TITLE
Post-training loss curves and training time estimates for OLMo-2 1B

### ISSUE THREAD

**Issue by BrownianNotion** [allenai/open-instruct#1296]
Hi team,
Would you be able to share the post-training loss curves and training time estimates for each post-training stage of the OLMo-2 1B model on an 8xH100 cluster? I cannot seem to find these in the OLMo-2 paper.

I've reviewed #485 and the OLMo-3 paper (which mentions a 9-day estimate), but both reference either larger models or more GPUs, so I'm hoping to get figures specific to this setup.

Thanks so much for any guidance!

**Comment by hamishivi** [comment]
Hi, 

SFT: 9 hours on one 1 H100 node (8 gpus).
DPO: 2 hours on one 1 H100 node (8 gpus).
RL: 2000000 episodes took roughly 43 hours on 16 H100s (2 nodes). We did two stages of RL training with same cost (so ~86 hours on 16 gpus all up).

Sharing the curves is a bit harder... I am not sure on the precise way to move out our wandb curves. Ping me again in the new year and I can see what I can do!

**Comment by BrownianNotion** [comment]
Thanks @hamishivi, will reach out in the new year about Wandb! DPO estimate lines up with that I've seen (perhaps closer to 1.5 hours with deepspeed stage 0). Haven't done any full SFT runs but here are some of my estimates on one 8xH100 node (based on tqdm and running ~200 steps):

- **default:** (`docs/olmo2.md`) 8+ hours (lines up with your estimate)
- **`per_device_batch_size=4`, `gradient_accumulation_steps=4`:** (same effective batch size) 5.5-6 hours
- **deepspeed stage 0:** 3.5-4 hours (`per_device_batch_size=2`)

It seems using deepspeed stage 0 instead of stage 2 gives huge time savings given that the model already fits in memory.

**Comment by BrownianNotion** [comment]
Hey @hamishivi, did you have any luck on sharing the wandb curves?

**Comment by hamishivi** [comment]
SFT: https://wandb.ai/ai2-llm/open_instruct_public/runs/mihc7mab
DPO: https://wandb.ai/ai2-llm/open_instruct_public/runs/k34ppe9q
The RL curves didnt transfer for some reason... working on it.

**Comment by BrownianNotion** [comment]
Hey @hamishivi , thanks so much! 
Any chance you know why the loss magnitudes are so different? Green is olmo, blue is my repro. Was the loss function, reduction method etc. changed? It's still a sum loss on `finetune.py` as far as I can see though.

<img width="537" height="303" alt="Image" src="https://github.com/user-attachments/assets/0f53e2d4-4037-4bcf-95ee-6e4ceb9b423a" />

<img width="549" height="314" alt="Image" src="https://github.com/user-attachments/assets/2a1e3155-1d81-45d1-898d-50cce585675f" />

**Comment by hamishivi** [comment]
Hi @BrownianNotion , this would be because we adjusted the metrics logging in the SFT script to normalize by the number of tokens, which makes things more comparable when changing the number of gpus used to train: https://github.com/allenai/open-instruct/blob/main/open_instruct/finetune.py#L860-L861

**Comment by hamishivi** [comment]
Unfortunately, I wasn't able to copy over the wandb curves for the RL jobs, it just didn't work. Here is some curves from the first RL job:

<img width="1464" height="713" alt="Image" src="https://github.com/user-attachments/assets/f2ff0c58-6f2e-4a25-9484-c1ba2cb76dda" />

And for the second:

<img width="1467" height="714" alt="Image" src="https://github.com/user-attachments/assets/c4c223b2-f338-47bc-a544-08633ea54aba" />

Note for this second run we only used MATH data so the reward shown is just the MATH reward.

Let me know if seeing a particular curve would be useful, sorry for the inconvenience!

**Comment by BrownianNotion** [comment]
@hamishivi thank you so much, these are still super helpful! If it's not too much trouble, I'd also love to see the loss curves (total, policy kl) and advantages (min, max, hist, val). Once again, really appreciate the help!

**Comment by hamishivi** [comment]
Here's the KL and loss curves! Unfortunately we did these runs before we had advantage logging.
Generally, I would expect loss to be flat and kl to slowly increase over time.


Run 1:

<img width="1462" height="514" alt="Image" src="https://github.com/user-attachments/assets/b0d69533-0032-45eb-8b73-6e8fc23cba39" />



Run 2:

<img width="1465" height="543" alt="Image" src="https://github.com/user-attachments/assets/4fc82406-2e25-4e7e-a068-e044ab6c6e95" />

**Comment by BrownianNotion** [comment]
Hey @hamishivi, I ran a few steps of the RLVR stage 1 training, but am getting very different loss magnitudes to your graphs. Did you also change the loss scaling/normalisation from the older script `grpo_vllm_thread_ray_gtrl.py` to the newer `grpo_fast.py`? This was run on an older checkout of the repo 5a2407345dea38937d2d5753cadff21c5e8777c8, before the Olmo core refactor of `grpo_fast.py`. 

<img width="1668" height="360" alt="Image" src="https://github.com/user-attachments/assets/eef6c751-aa7c-4b5e-b0cf-b06ffd955550" />

<img width="1098" height="303" alt="Image" src="https://github.com/user-attachments/assets/38a2e9b9-c564-4d57-a0a1-ccc875748dad" />

<img width="1099" height="309" alt="Image" src="https://github.com/user-attachments/assets/e9379624-e8cf-4d3c-b21b-7471e670b763" />


The command I ran was
```bash
python open_instruct/grpo_fast.py \
    --add_bos \
    --apply_verifiable_reward true \
    --beta 0.01 \
    --chat_template_name tulu \
    --dataset_cache_mode hf \
    --dataset_config_hash 7691399bb0 \
    --dataset_config_eval_hash 24054476c2 \
    --dataset_mixer_eval_list allenai/RLVR-GSM-MATH-IF-Mixed-Constraints 16 \
    --dataset_mixer_eval_list_splits train \
    --dataset_mixer_list allenai/RLVR-GSM-MATH-IF-Mixed-Constraints 1.0 \
    --dataset_mixer_list_splits train \
    --deepspeed_stage 2 \
    --deepspeed_zpg 12 \
    --exp_name 0423_grpo_seed_1_lr_5e-7_test_single_node \
    --fused_optimizer True \
    --kl_estimator 3 \
    --learning_rate 5e-7 \
    --local_eval_every 5 \
    --lr_scheduler_type constant \
    --max_prompt_token_length 2048 \
    --model_name_or_path allenai/OLMo-2-0425-1B-DPO \
    --model_revision main \
    --non_stop_penalty \
    --non_stop_penalty_value 0.0 \
    --num_epochs 1 \
    --num_learners_per_node 4 8 \
    --num_mini_batches 2 \
    --num_samples_per_prompt_rollout 16 \
    --num_unique_prompts_rollout 48 \
    --pack_length 4096 \
    --per_device_train_batch_size 1 \
    --remap_verifier ifeval=ifeval_old \
    --response_length 2048 \
    --save_freq 200 \
    --save_traces True \
    --seed 1 \
    --stop_strings "</answer>" \
    --temperature 1.0 \
    --tokenizer_name_or_path allenai/OLMo-2-0425-1B-DPO \
    --tokenizer_revision main \
    --total_episodes 2000000 \
    --verbose False \
    --vllm_gpu_memory_utilization 0.9 \
    --vllm_num_engines 4 \
    --vllm_sync_backend nccl \
    --vllm_tensor_parallel_size 1 \
    --with_tracking
```

**Comment by hamishivi** [comment]
@BrownianNotion I think the loss seems okay? The KL stuff I think we did change the way we computed the KL avg but not sure... main should be working okay these days!

Generally, I would just check to see if you see similar reward curves, and also just test downstream eval vs the released checkpoints.

**Comment by yukiwukii** [comment]
> Hey [@hamishivi](https://github.com/hamishivi) , thanks so much! Any chance you know why the loss magnitudes are so different? Green is olmo, blue is my repro. Was the loss function, reduction method etc. changed? It's still a sum loss on `finetune.py` as far as I can see though.
> 
> <img alt="Image" width="537" height="303" src="https://private-user-images.githubusercontent.com/49274027/533835547-0f53e2d4-4037-4bcf-95ee-6e4ceb9b423a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzM3NDEzNjEsIm5iZiI6MTc3Mzc0MTA2MSwicGF0aCI6Ii80OTI3NDAyNy81MzM4MzU1NDctMGY1M2UyZDQtNDAzNy00YmNmLTk1ZWUtNmU0Y2ViOWI0MjNhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMzE3VDA5NTEwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZlODRlOTljNGI3NThhYTQ4MDAxNmU0YmI4OGQ2ZWU1Njc4YTA3NDgwZjc0OTNhYjQwNDM2NTYxYzg3YjViZGMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._PN47Dz7uvK3QPQPw2ORTeqD6Ecwp0jMSpxjoWN5ENo"> <img alt="Image" width="549" height="314" src="https://private-user-images.githubusercontent.com/49274027/533835426-2a1e3155-1d81-45d1-898d-50cce585675f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzM3NDEzNjEsIm5iZiI6MTc3Mzc0MTA2MSwicGF0aCI6Ii80OTI3NDAyNy81MzM4MzU0MjYtMmExZTMxNTUtMWQ4MS00NWQxLTg5OGQtNTBjY2U1ODU2NzVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAzMTclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMzE3VDA5NTEwMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIzNTQ3MGFlMGI0MzU5MDliNDAxOWEzYzA4YjkxNTVjYTc0NjdjMDY2MTlhYzYzOTJjNzU3YWU1MGEzNGUzMWUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.hfGK16HbPWwODyEhJUtj4ZCHWs54A-9RmcIxH16CdnQ">

Hey @BrownianNotion, I am trying to reproduce OLMo2-1B-SFT as well. My train loss curve is similar to yours. However, when I tested my trained model on downstream tasks, its performance is significantly worse than the original SFT model. Did you experience the same thing?

This is similar to issue #751, but in my custom eval, the difference between my trained model and the checkpoint is large (~20% point).

**Comment by BrownianNotion** [comment]
Hey @yukiwukii , I'm not sure what your custom eval is, but there was indeed a significant difference between the OLMo-2 1B SFT checkpoint on hugging face and my attempt at reproducing it from performing SFT on the base model. See my comment here https://github.com/allenai/open-instruct/issues/1352#issuecomment-3804432037, there's about a 6% difference on GSM8K for example. If it's helpful to crosscheck, you can find my attempt at reconciling the reproduction commands [here](https://github.com/allenai/open-instruct/pull/1318).

**Comment by BrownianNotion** [comment]
Btw @hamishivi, the older version of `grpo_fast.py` (i.e. pre OLMo-core refactor) ran really fast even without `--inflight updates`! About 12 hours per stage for RLVR on 2x8x**H200**, so totalling around a day.

**Comment by hamishivi** [comment]
Oh great to know! I think the smaller models should be pretty fast. As for 1B SFT differences, I'm not sure. I think we did SFT training in olmo-core, so it might be some differences in training there (e.g., olmo core may have different setups around gradient norm).

**Comment by BrownianNotion** [comment]
@hamishivi feel free to close this issue, or leave it up for visibility. Thanks again for your help!
